codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/GoWriter.java

```diff
@@ -50,31 +55,236 @@
+        return (GoWriter w) -> {
+            w.writeGoBlockTemplate(beforeNewLine, afterNewLine, null, fn);
+        };
+    }
+
+    /**
+     * Returns a Writable that does nothing.
+     *
+     * @return Writable that does nothing
+     */
+    public static Writable emptyGoTemplate() {
+        return (GoWriter w) -> {
+        };
+    }
+
+    /**
+     * Writes the contents and arguments as a template to the writer.
+     *
+     * @param contents string to write
+     * @param args     Arguments to use when evaluating the contents string.
+     */
+    public void writeGoTemplate(String contents, Map<String, Object> args) {
+        withTemplate(contents, args, (template) -> {
+            try {
+                write(contents);
+            } catch (Exception e) {
+                throw new CodegenException("Failed to render template\n" + contents + "\nReason: " + e.getMessage(), e);
+            }
+        });
+    }
+
+    /**
+     * Writes the contents as template as a code block instead of single content fo text.
+     *
+     * @param beforeNewLine text before new line
+     * @param afterNewLine  text after new line
+     * @param args          template arguments
+     * @param fn            closure to write
+     */
+    public void writeGoBlockTemplate(
+            String beforeNewLine,
+            String afterNewLine,
+            Map<String, Object> args,
+            Consumer<GoWriter> fn
+    ) {
+        withTemplate(beforeNewLine, args, (header) -> {
+            conditionalBlock(header, afterNewLine, true, null, fn);
+        });
+    }
+
+    private void withTemplate(
+            String template,
+            Map<String, Object> scope,
+            Consumer<String> fn
+    ) {
+        if (scope == null) {
+            scope = new HashMap<>();
+        }
+
+        pushState();
+        putContext(scope);
+        validateContext(template, scope);
+        fn.accept(template);
+        popState();
+    }
+
+    private GoWriter conditionalBlock(
+            String beforeNewLine,
+            String afterNewLine,
+            boolean conditional,
+            Object[] args,
+            Consumer<GoWriter> fn
+    ) {
+        if (args == null) {
+            args = new Object[]{};
+        }
+        if (conditional) {
+            openBlock(beforeNewLine.trim(), args);
+        }
+
+        fn.accept(this);
+
+        if (conditional) {
+            closeBlock(afterNewLine.trim());
+        }
+
+        return this;
+    }
+
+    private static void validateTemplateArgsNotNull(Map<String, Object> scope) {
+        if (scope == null) {
+            return;
+        }
+        scope.forEach((k, v) -> {
+            if (v == null) {
+                throw new CodegenException("Template argument " + k + " cannot be null");
+            }
+        });
+    }
+
+    private void validateContext(String template, Map<String, Object> scope) {
+        var pattern = Pattern.compile("\\$([a-z][a-zA-Z_0-9]+)(:\\w)?");
```

Comment:
I'm not sure if Java would optimize this by default via static analysis, but can probably move this to be a `private final static` class field?